roachtest: grafana annotations read creds from cloud storage

DarrylWong opened the pull request:
As of https://github.com/cockroachdb/cockroach/pull/124099 we now store the service account creds in cloud storage. We already use this to access prometheus when generating dynamic configs. This change does the same for Grafana annotations by extracting the common logic into a helper.

This will allow users to have access to Grafana annotations out of the box locally, and limit the amount of benign but potentially confusing warnings about invalid credentials.

Release note: none
Fixes: none
Epic: none

DarrylWong committed:
roachtest: grafana annotations read creds from cloud storage

As of #124099 we now store the service account creds in cloud
storage. We already use this to access prometheus when generating
dynamic configs. This change does the same for Grafana annotations
by extracting the common logic into a helper.

This will allow users to have access to Grafana annotations out
of the box locally, and limit the amount of benign but potentially
confusing warnings about invalid credentials.

Epic: none
Fixes: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125571)
<!-- Reviewable:end -->

DarrylWong commented:
@nameisbhaskar I cleaned up some of the logging and changed the error handling a bit in `SetPromHelperCredsEnv`. Basically it didn't make sense to me to continue in the case of missing credentials since the oauth token retrieval will fail that case? But lmk if that seems sound to you.

renatolabs commented on pkg/cmd/roachprod/grafana/annotations.go:
```diff
@@ -42,25 +39,15 @@ func newGrafanaClient(
-		grafanaAudience := os.Getenv(ServiceAccountAudience)
-		if grafanaAudience == "" {
-			return nil, errors.Newf("%s env variable was not found", ServiceAccountAudience)
+		if _, err := promhelperclient.SetPromHelperCredsEnv(ctx, false); err != nil {
```
Documentation for `newGrafanaClient` could be updated to mention the function that reads credentials from env (maybe `SetPromHelperCredsEnv`), as `ServiceAccountJson` no longer exists.

renatolabs replied:
Also, the code reuse is good, but the fact that Grafana authentication works by calling `promhelperclient.SetPromHelperCredsEnv` is confusing because it suggests `promhelperclient` is involved.

renatolabs commented on pkg/roachprod/promhelperclient/client.go:
```diff
@@ -37,8 +36,8 @@ import (
-	serviceAccountJson     = "PROM_HELPER_SERVICE_ACCOUNT_JSON"
-	serviceAccountAudience = "PROM_HELPER_SERVICE_ACCOUNT_AUDIENCE"
+	ServiceAccountJson     = "PROM_HELPER_SERVICE_ACCOUNT_JSON"
+	ServiceAccountAudience = "PROM_HELPER_SERVICE_ACCOUNT_AUDIENCE"
```
Should these have more generic names? It's strange to me that we'll need to say that in order to use `roachprod grafana-annotation`, you'll need to set the `PROM_HELPER_SERVICE_ACCOUNT_JSON` env var.

renatolabs reviewed (approved):
LGTM in terms of code reuse; just have some naming questions.

> This will allow users to have access to Grafana annotations out of the box locally

I'm missing something in my understanding of how things are set up -- why does this change imply that everyone will have access out of the box? Doesn't it still require users to set environment variables locally?

DarrylWong commented:
> Doesn't it still require users to set environment variables locally?

`SetPromHelperCredsEnv` does the following in order until one succeeds:

1. Read in the env vars
2. Check if `promCredFile` exists
3. Read from cloud storage

In the case of number 3, it will set the environment variables locally after reading. It's a bit weird to me that the function doesn't just return the credentials and we need to refetch from the env, but I kept it as is to minimize changes.

DarrylWong commented:
I've extracted the common logic into a `roachprodutils` package and renamed them all to be generic to google idp so hopefully that helps remove the confusion.

DarrylWong commented:
TFTRs!

bors r=herkolategan, renatolabs
